### PR TITLE
Surface script from bitbucket

### PR DESCRIPTION
**Pull request by StongeEtienne**
Scripts to convert surface format, flip (from RAS to LPS), transform with ants warp, and smooth surfaces

https://usherbrooke-my.sharepoint.com/:u:/g/personal/stoe2204_usherbrooke_ca/ETLLgy0uF-dMk5bMB4F1aJEBFlF88_bhavmLgT_NOqKFPQ?e=VfR9Pf

example of use:
scil_flip_surface.py lh_white.vtk lh_white_lps.vtk -x -y
scil_transform_surface.py lh_white_lps.vtk transfo/vtk_affine.txt lh_white_lps_b0.vtk --ants_warp transfo/output1InverseWarp.nii.gz
scil_smooth_surface.py lh_white_lps_b0.vtk lh_white_lps_b0_smooth.vtk -m label/lh_mask.npy -n 5 -s 2.0 -v

### CODE REVIEW

**Comment by StongeEtienne** [comment]
I forgot to add that "pip install trimeshpy" should work, tested on fresh new "scilpy python 3.6 environments"

File: scripts/scil_flip_surface.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                   help='If supplied, flip the z axis.')
+
+    p.add_argument('-n', '--inverse_normal', action='store_true',
+                   help='If supplied, inverse surface orientation.')
+
+    add_overwrite_arg(p)
+    return p
+
+
+def main():
```
Comment (boiteaclou) [inline scripts/scil_flip_surface.py]: Deprecation warning in numpy_support

File: scripts/scil_flip_surface.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    assert_outputs_exist(parser, args, args.out_surface)
+
+    if not args.x and not args.y and not args.z and not args.inverse_normal:
+        parser.error('No action specified (flipping axis or inverse normal).')
+
+    # Load mesh
+    mesh = load_mesh_from_file(args.surface)
+
+    # Flip axes
+    flip = [-1 if args.x else 1,
```
Comment (boiteaclou) [inline scripts/scil_flip_surface.py]: This could be replaced with a store_const and a default on the arguments
`
parser.add_argument("--foo", action="store_const", const=-1, default=1, type=int)
`

**Comment by mdesco** [comment]
@StongeEtienne can you push your requirements with trimeshpy dependency

Also, please change scikit-learn==0.19.0 to scikit-learn==0.0

Thanks

File: scripts/scil_transform_surface.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                   help='Input surface (FreeSurfer or supported by VTK).')
+
+    p.add_argument('ants_affine',
+                   help='Affine transform from ANTs (.txt).')
+
+    p.add_argument('out_surface',
+                   help='Output surface (formats supported by VTK).')
+
+    p.add_argument('--ants_warp',
+                   help='Warp image from ANTs.')
```
Comment (boiteaclou) [inline scripts/scil_transform_surface.py]: Specify file format

File: scripts/scil_transform_surface.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+    assert_inputs_exist(parser, [args.surface, args.ants_affine],
+                        args.ants_warp)
+    assert_outputs_exist(parser, args, args.out_surface)
+
+    # Load mesh
+    mesh = load_mesh_from_file(args.surface)
+
+    # Affine transformation
+    if args.ants_affine is not None:
```
Comment (boiteaclou) [inline scripts/scil_transform_surface.py]: Can use only
`
if  not args.ants_affine
`

if None is always equal to False

File: scripts/scil_transform_surface.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        inv_affine = np.linalg.inv(affine)
+
+        # Transform mesh vertices
+        mesh.set_vertices(mesh.vertices_affine(inv_affine))
+
+        # Flip triangle face, if needed
+        if mesh.is_transformation_flip(inv_affine):
+            mesh.set_triangles(mesh.triangles_face_flip())
+
+    if args.ants_warp is not None:
```
Comment (boiteaclou) [inline scripts/scil_transform_surface.py]: Same here

File: scripts/scil_transform_surface.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        # Load warp
+        warp_img = nib.load(args.ants_warp)
+        warp = np.squeeze(warp_img.get_data())
```
Comment (boiteaclou) [inline scripts/scil_transform_surface.py]: Replace by `warp_img.get_fdata()`

File: scripts/scil_flip_surface.py (hunk 1 above)
Comment (StongeEtienne) [reply scripts/scil_flip_surface.py]: Comming from VTK  numpy support, I can't do much about it

File: scripts/scil_transform_surface.py (hunk 3 above)
Comment (StongeEtienne) [reply scripts/scil_transform_surface.py]: done

File: scripts/scil_transform_surface.py (hunk 4 above)
Comment (StongeEtienne) [reply scripts/scil_transform_surface.py]: changed to "if args.ants_affine:"

File: scripts/scil_transform_surface.py (hunk 6 above)
Comment (StongeEtienne) [reply scripts/scil_transform_surface.py]: done

File: scripts/scil_transform_surface.py (hunk 5 above)
Comment (StongeEtienne) [reply scripts/scil_transform_surface.py]: done same

**Comment by StongeEtienne** [comment]
@mdesco 
I think both should be done in a future PR since my trimeshpy version is not yet "fixed".
maybe I can start a PR "update denpendencies / requirements" 

> @StongeEtienne can you push your requirements with trimeshpy dependency
> 
> Also, please change scikit-learn==0.19.0 to scikit-learn==0.22.1
> 
> Thanks

File: scripts/scil_flip_surface.py (hunk 2 above)
Comment (StongeEtienne) [reply scripts/scil_flip_surface.py]: done

**Comment by StongeEtienne** [comment]
Updated  requirements : 
https://pypi.org/project/trimeshpy/0.0.2/

File: scripts/scil_flip_surface.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    p = argparse.ArgumentParser(description=__doc__, epilog=EPILOG,
+                                formatter_class=argparse.RawTextHelpFormatter)
+
+    p.add_argument('surface',
+                   help='Input surface (FreeSurfer or supported by VTK).')
+
+    p.add_argument('out_surface',
+                   help='Output flipped surface (formats supported by VTK).')
+
+    p.add_argument('-x', action="store_const", const=-1, default=1,
```
Comment (frheault) [inline scripts/scil_flip_surface.py:38]: Our 3 flipping scripts now use the convention.
```
    p.add_argument('axes',
                   choices=['x', 'y', 'z'], nargs='+',
                   help='The axes you want to flip. eg: to flip the x '
                        'and y axes use: x y.')
```

I think you should use that for consistency. I think your line `flip = [args.x, args.y, args.z]` would be fairly easy to adapt

File: scripts/scil_convert_surface.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+                   help='Output flipped surface (formats supported by VTK).')
+
+    add_overwrite_arg(p)
+    return p
+
+
+def main():
+    parser = _build_args_parser()
+    args = parser.parse_args()
+
```
Comment (frheault) [inline scripts/scil_convert_surface.py:42]: Do you think a check if the input extension is different from the output extension are different would be useful ?

File: scripts/scil_convert_surface.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+"""
+Script to convert cortical surface (vtk or freesurfer).
```
Comment (frheault) [inline scripts/scil_convert_surface.py:5]: Remove cortical (can be anything right?)
Can you list the accepted file format?

File: scripts/scil_flip_surface.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+"""
+Script to flip and reverse cortical surface (vtk or freesurfer).
```
Comment (frheault) [inline scripts/scil_flip_surface.py:5]: Remove cortical.
Maybe two sentences, one about flip and one about inverting normal just so it is clearer for the user.

Mentioning what is the use case for inverting the normal would be useful too,

File: scripts/scil_flip_surface.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+                             assert_outputs_exist)
+
+EPILOG = """
+References:
+[1] St-Onge, E., Daducci, A., Girard, G. and Descoteaux, M. 2018.
+    Surface-enhanced tractography (SET). NeuroImage.
+"""
+
+
+def _build_args_parser():
```
Comment (frheault) [inline scripts/scil_flip_surface.py:28]: _build_arg_parser

File: scripts/scil_convert_surface.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+                             assert_outputs_exist)
+
+EPILOG = """
+References:
+[1] St-Onge, E., Daducci, A., Girard, G. and Descoteaux, M. 2018.
+    Surface-enhanced tractography (SET). NeuroImage.
+"""
+
+
+def _build_args_parser():
```
Comment (frheault) [inline scripts/scil_convert_surface.py:25]: _build_arg_parser

File: scripts/scil_smooth_surface.py [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+
+EPILOG = """
+References:
+[1] St-Onge, E., Daducci, A., Girard, G. and Descoteaux, M. 2018.
+    Surface-enhanced tractography (SET). NeuroImage.
+"""
+
+
+def _build_args_parser():
```
Comment (frheault) [inline scripts/scil_smooth_surface.py:27]: _build_arg_parser

File: scripts/scil_smooth_surface.py [hunk 14]
```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+"""
+Script to smooth surface from a Laplacian blur.
```
Comment (frheault) [inline scripts/scil_smooth_surface.py:5]: This is a complex script, is there any corner case, recommendation, to do and not to do coming with it?
Like right now the default are made for a human cortical surface I assume ? It would be important to mention it.

You could also mention range of value if someone needs to fine-tune it a bit, for example I would have though that the parameters step_size would have been between 0.1 and 1 and nb_steps would have been around 20-50.
More details would be important for future user

File: scripts/scil_smooth_surface.py [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+    p.add_argument('out_surface',
+                   help='Output smoothed surface (formats supported by VTK).')
+
+    p.add_argument('-m', '--vts_mask',
+                   help='Vertices mask, where to apply the flow (.npy).')
+
+    p.add_argument('-n', '--nb_steps', type=int, default=2,
+                   help='Number of steps for laplacian smooth [%(default)s].')
+
+    p.add_argument('-s', '--step_size', type=float, default=5.0,
```
Comment (frheault) [inline scripts/scil_smooth_surface.py:43]: Is this in mm ? This script is likely made for RASMM or LPSMM for human so I assume it is, but I would mention it

File: scripts/scil_transform_surface.py [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+
+
+EPILOG = """
+References:
+[1] St-Onge, E., Daducci, A., Girard, G. and Descoteaux, M. 2018.
+    Surface-enhanced tractography (SET). NeuroImage.
+"""
+
+
+def _build_args_parser():
```
Comment (frheault) [inline scripts/scil_transform_surface.py:34]: _build_arg_parser

File: scripts/scil_transform_surface.py [hunk 17]
```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+"""
+Script to load and transform cortical surface (vtk or freesurfer),
+This script is using ANTs transform (affine.txt, warp.nii.gz).
+
+Best usage for with ANTs from T1 to b0:
+> ConvertTransformFile 3 output0GenericAffine.mat vtk_transfo.txt --hm
+> scil_transform_surface.py lh_white_lps.vtk affine.txt lh_white_b0.vtk\\
```
Comment (frheault) [inline scripts/scil_transform_surface.py:10]: From the filename I would think for that script to work "as intended" the input surface should be aligned with the T1 in MI-Brain and then the output would be aligned with the B0 in MI-Brain after.
Could you mention explicitly that the surface need to be Ips, or even simpler mention the MI-Brain sanity check so the user know in which 'space' the input must be.

File: scripts/scil_convert_surface.py [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+    p = argparse.ArgumentParser(description=__doc__, epilog=EPILOG,
+                                formatter_class=argparse.RawTextHelpFormatter)
+
+    p.add_argument('surface',
```
Comment (arnaudbore) [inline scripts/scil_convert_surface.py:29]: in_surface

File: scripts/scil_flip_surface.py [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+    p = argparse.ArgumentParser(description=__doc__, epilog=EPILOG,
+                                formatter_class=argparse.RawTextHelpFormatter)
+
+    p.add_argument('surface',
```
Comment (arnaudbore) [inline scripts/scil_flip_surface.py:32]: in_surface

File: scripts/scil_flip_surface.py (hunk 7 above)
Comment (arnaudbore) [reply scripts/scil_flip_surface.py:38]: Agree

File: scripts/scil_smooth_surface.py [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+[1] St-Onge, E., Daducci, A., Girard, G. and Descoteaux, M. 2018.
+    Surface-enhanced tractography (SET). NeuroImage.
+"""
+
+
+def _build_args_parser():
+    p = argparse.ArgumentParser(description=__doc__, epilog=EPILOG,
+                                formatter_class=argparse.RawTextHelpFormatter)
+
+    p.add_argument('surface',
```
Comment (arnaudbore) [inline scripts/scil_smooth_surface.py:31]: in_surface

File: scripts/scil_transform_surface.py [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+[1] St-Onge, E., Daducci, A., Girard, G. and Descoteaux, M. 2018.
+    Surface-enhanced tractography (SET). NeuroImage.
+"""
+
+
+def _build_args_parser():
+    p = argparse.ArgumentParser(description=__doc__, epilog=EPILOG,
+                                formatter_class=argparse.RawTextHelpFormatter)
+
+    p.add_argument('surface',
```
Comment (arnaudbore) [inline scripts/scil_transform_surface.py:38]: in_surface

File: scripts/scil_convert_surface.py (hunk 8 above)
Comment (StongeEtienne) [reply scripts/scil_convert_surface.py:42]: I don't really see the point, since there is no issues doing it

File: scripts/scil_smooth_surface.py [hunk 22]
```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+"""
+Script to smooth surface with a Laplacian blur.
+
+step_size in between 0.1 to 10 are recommended
```
Comment (frheault) [inline scripts/scil_smooth_surface.py:7]: step_size _in_ between 0.1 to 10 **is** recommended
(Not sure about the in between, your call)

File: scripts/scil_transform_surface.py [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+
+Best usage for with ANTs from T1 to b0:
+> ConvertTransformFile 3 output0GenericAffine.mat vtk_transfo.txt --hm
+> scil_transform_surface.py lh_white_lps.vtk affine.txt lh_white_b0.vtk\\
+    --ants_warp warp.nii.gz
+
+The input surface need to be in *T1 world LPS* coordinates
+(aligned over the T1 in MI-Brain).
+The resulting surface should be aligned *B0 world LPS* coordinates
+(aligned over the B0 in MI-Brain.
```
Comment (frheault) [inline scripts/scil_transform_surface.py:16]: Closing parenthesis

**Comment by arnaudbore** [comment]
@StongeEtienne tests are still missing.

File: scripts/scil_flip_surface.py [hunk 24]
```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+"""
+Script to flip and reverse surface (FreeSurfer or VTK supported).
+Can be used to flip in chosen axes (x, y or z),
+it can also flip inside out the surface orientation (normal).
```
[the remark this answers is not in the document]
Comment (StongeEtienne) [reply scripts/scil_flip_surface.py:7]: there is only one orientation in that case

File: scripts/scil_flip_surface.py [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+    Surface-enhanced tractography (SET). NeuroImage.
+"""
+
+
+def _build_arg_parser():
+    p = argparse.ArgumentParser(description=__doc__, epilog=EPILOG,
+                                formatter_class=argparse.RawTextHelpFormatter)
+
+    p.add_argument('in_surface',
+                   help='Input surface (FreeSurfer or supported by VTK).')
```
[the remark this answers is not in the document]
Comment (StongeEtienne) [reply scripts/scil_flip_surface.py:36]: I don't think I should list it, 
Like I already mention, 
1) it's purely base on VTK support and could change,
2) you should have used the "scil_convert_surface.py", maybe I should force ".vtk"

File: scripts/scil_smooth_surface.py [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+    Surface-enhanced tractography (SET). NeuroImage.
+"""
+
+
+def _build_arg_parser():
+    p = argparse.ArgumentParser(description=__doc__, epilog=EPILOG,
+                                formatter_class=argparse.RawTextHelpFormatter)
+
+    p.add_argument('in_surface',
+                   help='Input surface (FreeSurfer or supported by VTK).')
```
[the remark this answers is not in the document]
Comment (StongeEtienne) [reply scripts/scil_smooth_surface.py:38]: same as before comment

**Comment by mdesco** [comment]
FYI, my pip install -r requirements.txt error even with the master. So, it is not the fault of this PR. I think it is a python 3.7 problem that my virtualenv seem to get by default now instead of the python 3.5. 

@AntoineTheb did you have to do something special for tractometer?
thanks

Otherwise, I think I have to force a lower python version

File: scripts/scil_smooth_surface.py [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+
+    p.add_argument('out_surface',
+                   help='Output smoothed surface (formats supported by VTK).')
+
+    p.add_argument('-m', '--vts_mask',
+                   help='Vertices mask, where to apply the flow (.npy).')
+
+    p.add_argument('-n', '--nb_steps', type=int, default=2,
+                   help='Number of steps for laplacian smooth [%(default)s].')
```
Comment (mdesco) [inline scripts/scil_smooth_surface.py:47]: I think all our other scilpy scripts use [default: %(default)s].

Can you please add this here too.

Comment (StongeEtienne) [reply scripts/scil_smooth_surface.py:47]: I look in general and there is more [%(default)s] then  (Default: %(default)s)
recent script also use  [%(default)s] 
@arnaudbore  @frheault ?
